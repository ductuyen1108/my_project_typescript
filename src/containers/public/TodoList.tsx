import React, { useState, useRef } from 'react';
import { Alert, Box, Button, List, ListItem, TextField, Typography, Paper } from '@mui/material';
import { Delete, Edit, Save, Cancel } from '@mui/icons-material';
import { useDispatch, useSelector } from 'react-redux';
import { addTodo, deleteTodo, editTodo } from '../../slice/todoListSlice';
import { todoListSelector } from '../../redux/selectors';

interface Todo {
    id: string;
    job: string;
}

const TodoList: React.FC = () => {
    const [work, setWork] = useState<string>('');
    //const [todos, setTodos] = useState<Todo[]>([]);
    const [showAlert, setShowAlert] = useState<boolean>(false);
    const [editingTodoId, setEditingTodoId] = useState<string | null>(null);
    const [editedTodo, setEditedTodo] = useState<string>('');

    const inputRef = useRef<HTMLInputElement>(null);
    const dispatch = useDispatch();

    const todos = useSelector(todoListSelector);
    console.log(todos);

    /* useEffect(() => {
        const storedTodos = localStorage.getItem('todos');
        if (storedTodos) {
            setTodos(JSON.parse(storedTodos));
        }
    }, []); */

    /* useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]); */

    /* const handleAdd = () => {
        if (todos.some((todo) => todo.id === work.replace(/\s/g, ''))) {
            setShowAlert(true);
        } else {
            setTodos((prev) => [...prev, { id: work.replace(/\s/g, ''), job: work }]);
            setWork('');
            if (inputRef.current) {
                inputRef.current.focus();
            }
            setShowAlert(false);
        }
    }; */

    const handleAdd = () => {
        const id = work.replace(/\s/g, '');
        if (todos.some((todo) => todo.id === work.replace(/\s/g, ''))) {
            setShowAlert(true);
        } else {
            dispatch(addTodo({ id: id, job: work }));
            setWork('');
            if (inputRef.current) {
                inputRef.current.focus();
            }
            setShowAlert(false);
        }
    };

    /* const handleDelete = (id: string) => {
        setTodos((prev) => prev.filter((todo) => todo.id !== id));
    }; */

    const handleDelete = (id: string) => {
        dispatch(deleteTodo(id));
    };

    const handleEdit = (id: string, job: string) => {
        setEditingTodoId(id);
        setEditedTodo(job);
    };

    const handleSave = () => {
        /* setTodos((prevTodos) =>
            prevTodos.map((todo) => {
                if (todo.id === editingTodoId) {
                    return { ...todo, job: editedTodo };
                }
                return todo;
            }),
        ); */
        dispatch(editTodo({ id: editingTodoId!, job: editedTodo }));
        setEditingTodoId(null);
    };

    const handleCancelEdit = () => {
        setEditingTodoId(null);
        setEditedTodo('');
    };

    return (
        <Box
            sx={{
                display: 'flex',
                flexDirection: 'column',
                alignItems: 'center',
                textAlign: 'center',
                backgroundColor: '#f5f5f5',
            }}
        >
            <Paper
                sx={{
                    width: '800px',
                    height: '600px',
                    mt: '30px',
                    display: 'flex',
                    flexDirection: 'column',
                    alignItems: 'center',
                }}
            >
                <Box sx={{ width: '600px', display: 'flex', flexDirection: 'column', alignItems: 'center' }}>
                    <Typography sx={{ marginBottom: '10px', marginTop: '10px', fontSize: '35px' }}>
                        Todo List
                    </Typography>
                    <Box>
                        <Box
                            sx={{
                                display: 'flex',
                                gap: '35px',
                                height: '40px',
                                alignItems: 'center',
                                marginBottom: '20px',
                            }}
                        >
                            <TextField
                                //label="Enter todo"
                                placeholder="Enter todo"
                                variant="standard"
                                value={work}
                                ref={inputRef}
                                fullWidth
                                required
                                onChange={(e) => setWork(e.target.value)}
                            />
                            <Button variant="contained" sx={{ width: '150px' }} onClick={handleAdd}>
                                Add
                            </Button>
                        </Box>
                        <Box>
                            <List sx={{ width: '500px' }}>
                                {todos.map((todo) => (
                                    <ListItem
                                        key={todo.id}
                                        sx={{
                                            display: 'flex',
                                            width: '100%',
                                            justifyContent: 'space-between',
                                            alignItems: 'center',
                                        }}
                                    >
                                        <Typography sx={{ width: '220px' }}>
                                            {editingTodoId === todo.id ? (
                                                <TextField
                                                    value={editedTodo}
                                                    onChange={(e) => setEditedTodo(e.target.value)}
                                                    fullWidth
                                                    required
                                                    autoFocus
                                                />
                                            ) : (
                                                todo.job
                                            )}
                                        </Typography>
                                        {editingTodoId === todo.id ? (
                                            <Box display={'flex'} gap={2}>
                                                <Button
                                                    variant="outlined"
                                                    startIcon={<Save />}
                                                    onClick={handleSave}
                                                    disabled={!editedTodo.trim()}
                                                >
                                                    Save
                                                </Button>
                                                <Button
                                                    variant="outlined"
                                                    startIcon={<Cancel />}
                                                    onClick={handleCancelEdit}
                                                >
                                                    Cancel
                                                </Button>
                                            </Box>
                                        ) : (
                                            <Box display={'flex'} gap={2}>
                                                <Button
                                                    variant="outlined"
                                                    startIcon={<Edit />}
                                                    onClick={() => handleEdit(todo.id, todo.job)}
                                                >
                                                    Edit
                                                </Button>
                                                <Button
                                                    variant="outlined"
                                                    startIcon={<Delete />}
                                                    onClick={() => handleDelete(todo.id)}
                                                >
                                                    Delete
                                                </Button>
                                            </Box>
                                        )}
                                    </ListItem>
                                ))}
                            </List>
                        </Box>
                    </Box>
                    {showAlert && <Alert severity="error">Công việc đã thêm vào trước đó!</Alert>}
                </Box>
            </Paper>
        </Box>
    );
};

export default TodoList;
