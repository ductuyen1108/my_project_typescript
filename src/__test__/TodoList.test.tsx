/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TodoList from '../containers/public/TodoList';

describe('TodoList', () => {
    it('should match snapshot', () => {
        const { asFragment } = render(<TodoList />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should add a todo when clicking the Add button', () => {
        const { getByLabelText, getByText } = render(<TodoList />);
        const input = getByLabelText('Enter todo');
        const addButton = getByText('Add');

        fireEvent.change(input, { target: { value: 'New todo' } });
        fireEvent.click(addButton);

        const todoItem = getByText('New todo');
        expect(todoItem).toBeInTheDocument();
    });

    it('should edit a todo', () => {
        const { getByText, getByLabelText } = render(<TodoList />);
        const editButton = getByText('Edit');
        fireEvent.click(editButton);

        const editInput = getByLabelText('Enter todo');
        fireEvent.change(editInput, { target: { value: 'Edited todo' } });

        const saveButton = getByText('Save');
        fireEvent.click(saveButton);

        const editedTodo = getByText('Edited todo');
        expect(editedTodo).toBeInTheDocument();
    });

    it('should delete a todo', () => {
        const { getByText } = render(<TodoList />);
        const deleteButton = getByText('Delete');
        fireEvent.click(deleteButton);

        expect(deleteButton).not.toBeInTheDocument();
    });
});
