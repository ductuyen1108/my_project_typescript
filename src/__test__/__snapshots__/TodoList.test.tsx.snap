// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-uyhq4l"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1d1h7pn-MuiPaper-root"
    >
      <div
        class="MuiBox-root css-mghrfl"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-gsqi3c-MuiTypography-root"
        >
          Todo List
        </p>
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiBox-root css-hnk2cx"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                Enter todo
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                  â€‰*
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  id=":r0:"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1ak2d8a-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Add
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <ul
              class="MuiList-root MuiList-padding css-19u38oo-MuiList-root"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
